goaci.go

```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
-	if err != nil {
-		die("error opening output file: %v", err)
+		"-installsuffix", "nocgo", // for 1.4
+		opts.project,
 	}
 
 	env := []string{
-		"GOPATH=" + goPathOpt,
-		"GOBIN=" + gobin,
+		"GOPATH=" + pathsNames.goPath,
+		"GOBIN=" + pathsNames.goBinPath,
 		"CGO_ENABLED=0",
 		"PATH=" + os.Getenv("PATH"),
 	}
-	if goRoot != "" {
-		env = append(env, "GOROOT="+goRoot)
+	if pathsNames.goRootPath != "" {
+		env = append(env, "GOROOT="+pathsNames.goRootPath)
 	}
+
 	cmd := exec.Cmd{
 		Env:    env,
-		Path:   goBinaryOpt,
+		Path:   pathsNames.goExecPath,
 		Args:   args,
 		Stderr: os.Stderr,
 		Stdout: os.Stdout,
 	}
-	debug("env:", cmd.Env)
-	debug("running command:", strings.Join(cmd.Args, " "))
+	Debug("env: ", cmd.Env)
+	Debug("running command: ", strings.Join(cmd.Args, " "))
 	if err := cmd.Run(); err != nil {
-		die("error running go: %v", err)
+		return err
 	}
+	return nil
+}
 
-	// Check that we got 1 binary from the go get command
-	fi, err := ioutil.ReadDir(gobin)
+// getBinaryName get a binary name built by go get and selected by
+// --use-binary parameter.
+func getBinaryName(opts *options, pathsNames *pathsAndNames) (string, error) {
+	fi, err := ioutil.ReadDir(pathsNames.goBinPath)
 	if err != nil {
-		die(err.Error())
+		return "", err
 	}
+
 	switch {
 	case len(fi) < 1:
-		die("no binaries found in gobin")
+		return "", fmt.Errorf("No binaries found in gobin.")
+	case len(fi) == 1:
+		name := fi[0].Name()
+		if opts.useBinary != "" && name != opts.useBinary {
+			return "", fmt.Errorf("No such binary found in gobin: %s. There is only %s", opts.useBinary, name)
+		}
+		Debug("found binary: ", name)
+		return name, nil
 	case len(fi) > 1:
-		debug(fmt.Sprint(fi))
-		die("can't handle multiple binaries")
+		names := []string{}
+		for _, v := range fi {
+			names = append(names, v.Name())
+		}
+		if opts.useBinary == "" {
+			return "", fmt.Errorf("Found multiple binaries in gobin, but --use-binary option is not used. Please specify which binary to put in ACI. Following binaries are available: %s", strings.Join(names, ", "))
+		}
+		for _, v := range names {
+			if v == opts.useBinary {
+				return v, nil
+			}
+		}
+		return "", fmt.Errorf("No such binary found in gobin: %s. There are following binaries available: %s", opts.useBinary, strings.Join(names, ", "))
+	}
+	return "", fmt.Errorf("Reaching this point shouldn't be possible.")
+}
+
+func getApp(opts *options, binary string) *types.App {
+	exec := []string{filepath.Join("/", binary)}
+	exec = append(exec, opts.exec...)
+
+	return &types.App{
+		Exec:  exec,
+		User:  "0",
+		Group: "0",
 	}
-	fn := fi[0].Name()
-	debug("found binary: ", fn)
+}
 
-	// Set up rootfs for ACI layout
-	rfs := filepath.Join(acidir, "rootfs")
-	err = os.MkdirAll(rfs, 0755)
+func getVCSLabel(pathsNames *pathsAndNames) (*types.Label, error) {
+	name, value, err := GetVCSInfo(pathsNames.projectPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get VCS info: %v", err)
```

Comment:
I'm not sure I'd like to add support for that. That would probably need adding another option telling whether last parameter passed to goaci is a project name or a directory which it should try work on. It can get ambiguous in one corner case - you are in $GOPATH/src and call `goaci github.com/coreos/etcd`. Should `goaci` try to basically do `cd github.com/coreos/etcd` or rather run `go get github.com/coreos/etcd`?

But I'd rather file a separate issue about it.